File: src/test/java/com/enchainte/sdk/e2e/E2ETest.java
package com.enchainte.sdk.e2e;

import com.enchainte.sdk.EnchainteClient;
import com.enchainte.sdk.common.BaseTest;
import com.enchainte.sdk.message.domain.Message;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.Random;

import static org.junit.Assert.assertTrue;

public class E2ETest {
    @Test
    public void e2eJavaTest() throws Exception {
        Message message = Message.fromHex(getRandomHexString());

        EnchainteClient client = new EnchainteClient("xfJvoW4xZONxW2tDD5Vypwu3S23LfC-GaexO2epA7UjHilfFSm4gaMeZUuvcJtXz");
        System.out.println("SENDING MESSAGE: " + message.getHash());
        client.sendMessage(message).subscribe();

        System.out.println("WAITING MESSAGE");
        client.waitMessageReceipts(Collections.singletonList(message)).blockingSubscribe();

        System.out.println("VALIDATING MESSAGE");
        var valid = false;
        while (!valid) {
            valid = client.verifyMessages(Collections.singletonList(message)).blockingGet();
            Thread.sleep(500);
        }

        assertTrue(valid);
    }

    private String getRandomHexString(){
        Random r = new Random();
        StringBuilder sb = new StringBuilder();
        while(sb.length() < 32){
            sb.append(Integer.toHexString(r.nextInt()));
        }

        return sb.substring(0, 32);
    }
}

File: src/test/kotlin/com/enchainte/sdk/e2e/E2ETest.kt
package com.enchainte.sdk.e2e

import com.enchainte.sdk.EnchainteClient
import com.enchainte.sdk.common.BaseTest
import com.enchainte.sdk.message.domain.Message
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.Test
import kotlin.test.assertTrue

class E2ETest {
    @Test
    fun e2eTest() {
        val charPool : List<Char> = ('a'..'f') + ('0'..'9')
        val randomHex = (1..16)
            .map { _ -> kotlin.random.Random.nextInt(0, charPool.size) }
            .map(charPool::get)
            .joinToString("");

        val message = Message.fromHex(randomHex)

        val client = EnchainteClient("xfJvoW4xZONxW2tDD5Vypwu3S23LfC-GaexO2epA7UjHilfFSm4gaMeZUuvcJtXz")
        println("SENDING MESSAGE: ${message.getHash()}")
        client.sendMessage(message).subscribe()

        runBlocking {
            println("WAITING MESSAGE")
            client.waitMessageReceipts(listOf(message)).blockingSubscribe()

            println("VALIDATING MESSAGE")
            var valid = false
            while (!valid) {
                valid = client.verifyMessages(listOf(message)).blockingGet()
                delay(500)
            }

            assertTrue(valid)
        }
    }
}